rust-runtime/aws-smithy-async/src/future/pagination_stream.rs

```diff
@@ -3,16 +3,51 @@
  * SPDX-License-Identifier: Apache-2.0
  */
 
-//! Utility to drive a stream with an async function and a channel.
+//! Types to support stream-like operations for paginators.
 
 use crate::future::rendezvous;
-use futures_util::StreamExt;
 use pin_project_lite::pin_project;
+use std::fmt;
+use std::future::poll_fn;
 use std::future::Future;
 use std::pin::Pin;
 use std::task::{Context, Poll};
-use tokio_stream::{Iter, Once, Stream};
 
+pub mod collect;
+
+/// A wrapper around [`FnStream`].
+///
+/// This type provides the same set of methods as [`FnStream`], but that is meant to be used
+/// internally and not by external users.
+#[derive(Debug)]
+pub struct PaginationStream<Item>(FnStream<Item>);
+
+impl<Item> PaginationStream<Item> {
+    /// Creates a `PaginationStream` from the given [`FnStream`].
+    pub fn new(stream: FnStream<Item>) -> Self {
+        Self(stream)
+    }
+
+    /// Consumes and returns the next `Item` from this stream.
+    pub async fn next(&mut self) -> Option<Item>
+    where
+        Self: Unpin,
+    {
+        self.0.next().await
+    }
+
+    /// Consumes this stream and gathers elements into a collection.
+    pub async fn collect<T: collect::sealed::Collectable<Item>>(self) -> T {
+        self.0.collect().await
+    }
+}
```

Comment:
I guess you can't map streams anymore? We could add a conversion to `aws-smithy-types-convert` I guess to regain access to `StreamExt`?

Reply:
As we discussed, can be done in an additive manner outside of the PR. https://github.com/awslabs/smithy-rs/issues/2993 keeps track of it.